### DISCUSSION TITLE
Experimenting with Soroban Auth

### DISCUSSION THREAD

**Discussion by mootz12** [stellar/sorobanathon#30] [General]
Soroban provides a unique authentication system with built-in multi-sig and more. How does it hold up when implementing two notable authorization patterns found in Ethereum: `permit` ([Uniswap V3 Router](https://github.com/Uniswap/v3-periphery/blob/main/contracts/base/SelfPermit.sol)) and `delegate` ([Aave V3](https://docs.aave.com/developers/core-contracts/pool#borrow))?

**`Disclaimer`** The following code is not well tested and shouldn't be relied on for anything, as there are likely bugs.
Github repo containing the full codebase: https://github.com/mootz12/soroban-auth-pool

If there are other authorization situations you would like to see - include them in the comments and I'll try and get them added!

## Setup
Pools are a great environment for showcasing different ways of authorization. They hold tokens from a variety of sources, and have full control of how those tokens can be accessed.

The `soroban-auth-pool` in the example currently allows users to deposit a specific token into the pool, and the pool grants them *collateral* as credit for their deposit. The user can then delegate this collateral to another user, who can borrow tokens up to the delegator's allotted collateral.

Other functions like `repay` and `withdraw` are omitted from the contract, as they don't affect the two auth patterns of `permit` and `delegate`.

## Permit
Tokens in Soroban and other smart contract environments currently rely on an allowance system to help ensure contracts don't inappropriately spend funds. However, early iterations forced users to send an `approval` transaction, followed by another transaction to call the contract function they originally cared about. [EIP-2612 Permit](https://eips.ethereum.org/EIPS/eip-2612) was proposed as a fix to allow single transaction approvals.

### Implementation
The native token contract does not contain a specific `permit` function. However, the [advanced auth](https://soroban.stellar.org/docs/examples/auth-advanced) functionality can be utilized to pre-sign an `approval` function call. Thus, the pool can attempt to run `approval` on behalf of the user before it performs a `xfer_from`.

The `soroban-auth-pool` exposes a `deposit_p` method to deposit into the pool with an embedded approval:
```rust
/// Runs an approval and deposit in the same transcation
fn deposit_p(e: Env, token_approval_sig: Signature, amount: i64) {
    let sig_id = token_approval_sig.identifier(&e);
    let token_client = get_token_client(&e);
    let amount_bi = &BigInt::from_i64(&e, amount);

    // run the approval
    let sender_nonce = token_client.nonce(&sig_id);
    token_client.approve(&token_approval_sig, &sender_nonce, &get_contract_id(&e), &amount_bi);

    // now the pool has the appropriate permissions to run `transfer_from`
    token_client.xfer_from(
        &Signature::Invoker,
        &BigInt::zero(&e),
        &sig_id,
        &get_contract_id(&e),
        &BigInt::from_i64(&e, amount)
    );

    let cur_collateral = get_collateral(&e, sig_id.clone());
    set_collateral(&e, sig_id, cur_collateral + amount);
}
```
It's important to note that once this transaction reaches core nodes, anybody who is listening can attempt to submit this as nothing is derived from the invoker of the transaction. To ensure the transaction cannot be called twice, the transaction will `panic` if the `approve` method on the token fails. This means this function inherits the nonce benefits implemented by the native token contract.

### Result
Permit is achieved "for free" by Soroban's ability to pre-sign full transaction signatures for the native token contract, so no specific `permit` function is required.

One caveat about the lack of `permit` function is there is no effective deadline enforced by default. As such, if a user submits a transaction that fails and contains one of these pre-signed approvals, anyone listening for transactions can submit the pre-signed approval at a later date.

Lastly, this is almost implementable in a stateless format, where no approval tracking would be necessary. The only detractor is `xfer` does not include any way to validate who the `invoker` or `sender` is on a transaction, so a pre-signed transfer message from the user would not be safe to submit as calldata on chain. 

## Delegate
Delegation allows contracts to extend access to a specific benefit, like votes or collateral, to a different address. Oftentimes, it's implemented similarly to `permit` with specific data structures to store delegation ([OpenZeppelin Vote Example](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.7.3/contracts/governance/utils/Votes.sol#L92)).

### Implementation
The `soroban-auth-pool` adds *credit delegation*, which allows owners of collateral to delegate it out to specific users who can borrow from the pool. This allows (for a money market pool) market makers and arbitrageurs to generate large sums of collateral from a single address, and delegate it to specific trading contracts that need to borrow capital in an instant.

A signature is used to validate the conditions of the borrow: the address receiving the funds (the sender/invoker), the amount, and a deadline the borrow needs to be executed by. Thus, the signer is able to delegate collateral to the sender of the transaction to borrow funds:

```rust
/// A signature gives permission to the sender to borrow funds from the signer's collateral balance!
fn borrow_obo(e: Env, sig: Signature, amount: i64, expiration: u64) {
    // Verify signature is not expired
    if expiration < e.ledger().timestamp() {
        panic!("expired signature")
    }
    
    // Verify that the signature signs and authorizes this invocation.
    let signer_id = sig.identifier(&e);
    let sender_id = Identifier::from(e.invoker());
    let nonce = get_nonce(&e, &signer_id);

    // By including the `sender` in the signature alongside the `contract_id` and function `symbol`
    // the signer can be ensured nobody other than the sender can execute against this signature
    verify(&e, &sig, symbol!("borrow_obo"), (&signer_id, &nonce, &sender_id, &amount, &expiration));
    verify_and_consume_nonce(&e, &sig, &nonce);

    // Check collateral and liability balances
    let signer_collateral = get_collateral(&e, signer_id.clone());
    let signer_liability = get_liabilities(&e, signer_id.clone());
    if signer_collateral < (signer_liability + amount) {
        panic!("not enough collateral")
    }

    // Looks good - execute token distribution
    set_liabilities(&e, signer_id, signer_liability + amount);

    let token_client = get_token_client(&e);
    token_client.xfer(&Signature::Invoker, &BigInt::zero(&e), &sender_id, &BigInt::from_i64(&e, amount));
}
```

### Result
Credit delegation was achieved without requiring any additional data structures to track delegation. However, the delegator is required to sign a message each time access is delegated. This type of delegation works best for *one-to-many*, where one address delegates access to others. For a use case like voting delegation, which is *many-to-one*, forcing the one to organize signatures from multiple parties would create lots of friction. Following the `permit` pattern shown by the OpenZeppelin Votes contract would be a better solution.

**Comment by achallagundla** [comment]
Heya @mootz12  thank you for your submission...will get back to you shortly after review!

**Comment by achallagundla** [comment]
Heya @mootz12 ,
Congratulations! Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of **2500 XLM**! One last step: to receive that reward, please fill out this [form](https://stellarform.typeform.com/Sorobanathon) completely and accurately. 

This post really helped our Soroban Dev team! Thank you! 

Please let us know if you have any questions. (You can also message me on discord @ anuxhya)

All the best,
Anuhya
